### PR TITLE
Verify the quoting enclave report body

### PR DESCRIPTION
**Pull request by nick-mobilecoin**
Add logic to verify the quoting enclave report body from a quote.

This only contains the logic to verify the quote body.  The other steps
necessary to fulfill the quote verification are listed in the issue #17.

There is an [okteta](https://apps.kde.org/okteta/) definition available at
https://github.com/nick-mobilecoin/okteta-SGXQuote to make it easier for one to
view the `hw_quote.dat` file.  I'm up for suggestions on how to name the current
`hw_quote.dat` file.

### CODE REVIEW

File: untrusted/.gitignore [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Generated by Cargo
+# will have compiled files and executables
+/target/
+
+# Remove Cargo.lock from gitignore if creating an executable, leave it for libraries
+# More information here https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html
+Cargo.lock
+
+# These are backup files generated by rustfmt
+**/*.rs.bk
```
Comment (remoun) [inline untrusted/.gitignore:10]: Optional/Equivalent: Omit the `**/`
```suggestion
*.rs.bk
```

File: untrusted/quote_verify/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+// ASN.1 Tag for a sequence
+const ASN1_SEQUENCE: u8 = 48;
+
+/// A quote for DCAP attestation
+pub struct Quote {
+    bytes: Vec<u8>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
```
Comment (remoun) [inline untrusted/quote_verify/src/lib.rs:41]: Nit: Move Error to its own file or at least below `impl Quote`

File: untrusted/quote_verify/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    /// Returns the quoting enclave report from the overall quote.
+    fn get_quoting_enclave_report(&self) -> &[u8] {
+        &self.bytes
+            [QUOTING_ENCLAVE_REPORT_START..QUOTING_ENCLAVE_REPORT_START + ENCLAVE_REPORT_SIZE]
+    }
+
+    /// Returns the ASN.1 version of the signature.
+    /// mbedtls wants an ASN.1 version of the signature, while the raw quote
+    /// format has only the r and s values of the ECDSA signature
+    fn get_ans1_signature(&self) -> Vec<u8> {
```
Comment (remoun) [inline untrusted/quote_verify/src/lib.rs:101]: Typo, should be `asn1`

File: untrusted/quote_verify/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+
+use mbedtls::alloc::Box as MbedtlsBox;
+use mbedtls::hash::Type as HashType;
+use mbedtls::x509::Certificate;
```
Comment (remoun) [inline untrusted/quote_verify/src/lib.rs:5]: Prefer grouped imports

File: untrusted/quote_verify/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    /// mbedtls wants an ASN.1 version of the signature, while the raw quote
+    /// format has only the r and s values of the ECDSA signature
+    fn get_ans1_signature(&self) -> Vec<u8> {
+        let mut start = QUOTING_ENCLAVE_SIGNATURE_START;
+        let r = &self.bytes[start..start + SIGNATURE_COMPONENT_SIZE];
+        start += SIGNATURE_COMPONENT_SIZE;
+        let s = &self.bytes[start..start + SIGNATURE_COMPONENT_SIZE];
+
+        // The `2`s are the tag byte + length byte
+        let length = 2 + r.len() + 2 + s.len();
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply untrusted/quote_verify/src/lib.rs]: Done.
Note the `usize`, due to how it's used in the addition there with `.len()`

File: untrusted/quote_verify/src/lib.rs (hunk 2 above)
Comment (nick-mobilecoin) [reply untrusted/quote_verify/src/lib.rs]: Moved Error below
My initial thought was to leave Error in this module until #15 is able to start coming in, but perhaps I  should visit this sooner than later.

File: untrusted/quote_verify/src/lib.rs (hunk 3 above)
Comment (nick-mobilecoin) [reply untrusted/quote_verify/src/lib.rs]: good catch!  I kept googling ans1 and was getting a bunch of ansi stuff :)

File: untrusted/quote_verify/src/lib.rs (hunk 4 above)
Comment (nick-mobilecoin) [reply untrusted/quote_verify/src/lib.rs]: doh, I forget about doing this for nested imports, thanks

**Review by remoun — APPROVED** [review]

File: untrusted/quote_verify/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        let s = &self.bytes[start..start + SIGNATURE_COMPONENT_SIZE];
+
+        let sequence_length = ASN1_TYPE_LENGTH_SIZE + r.len() + ASN1_TYPE_LENGTH_SIZE + s.len();
+        let mut signature = vec![ASN1_SEQUENCE, sequence_length as u8];
+        for component in [r, s] {
+            signature.extend([ASN1_INTEGER, component.len() as u8]);
+            signature.extend(component);
+        }
+        signature
+    }
```
Comment (jcape) [inline untrusted/quote_verify/src/lib.rs:100]: This is actually a DER-formatted signature.

File: untrusted/quote_verify/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    ///
+    /// # Arguments
+    ///
+    /// * `bytes` the bytes of the quote as defined in
+    ///     https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf
+    pub fn from_bytes(bytes: &[u8]) -> Self {
+        Quote {
+            bytes: bytes.to_vec(),
+        }
+    }
```
Comment (jcape) [inline untrusted/quote_verify/src/lib.rs:53]: Is there any advantage to not making this fallible and decomposing the bytes into a real structure?

**Pull request merged**